### ISSUE TITLE
Potential approaches for performance improvements in `cftime.num2date`

### ISSUE THREAD

**Issue by spencerkclark** [Unidata/cftime#269]
### Issue

While debugging a user issue, a performance limitation of `cftime.num2date` was put in sharp relief in pydata/xarray#6284.  Essentially `num2date` gets slower the farther the encoded times are from their reference date.  We sort of knew about this when we adopted the new decoding algorithm in #176, but it didn't seem that serious until we encountered an array with 15+ million times.  Arrays of times that large are admittedly rare, but it could be good to explore some options for improving the performance of `num2date`, since, as suggested by @rabernat and as I show below, there is some room for improvement.

A minimal example that illustrates this performance issue is the following:
```
In [1]: import cftime; import numpy as np

In [2]: encoded_times = 365 * 2000 + np.linspace(0, 30, 10000)

In [3]: %timeit cftime.num2date(encoded_times, "days since 0001-01-01", "proleptic_gregorian")
3.21 s ± 34.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

If we bring the reference date closer to the encoded times, we obtain an over 250x speedup.

```
In [4]: encoded_times = np.linspace(0, 30, 10000)

In [5]: %timeit cftime.num2date(encoded_times, "days since 1999-09-04", "proleptic_gregorian")
11.5 ms ± 142 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

This example uses 10,000 times, and even there 3 seconds is slow.  With 15 million times, it would take over an hour.  With a better reference date, however, 15 million times would take about 15 seconds (still slow, but bearable).  

### Possible solutions

@rabernat suggested two potential performance optimizations that seem fairly natural to try:

> One possible performance optimization you could make in cftime.num2date would be to use the true reference date just for the first value of the times arrray. Then you could say times_delta = times - times[0] (re-reference everything to the first timestep) and update the reference date accordingly. Going even further, you could imagine computing each time relative as a timedelta relative to the previous timestep.

### Results

I implemented these locally -- along with a couple more possible optimizations -- and did some profiling.  There are a few use-cases to consider when profiling.  For each of these, to emphasize the original issue, I used a distant reference date.   I implemented these in [a branch](https://github.com/spencerkclark/cftime/tree/num2date-optimization) and ran the profiling script in [this gist](https://gist.github.com/spencerkclark/84e552bafd942fd6f7df73b253e63002) to generate these results (note for simplicity for now I did not cover all edge cases involving zero-size or masked arrays).  The example above corresponds with the "Ordered narrow" case.  These are all for 10,000 times.

|                                                       | Ordered narrow | Ordered long | Unordered narrow | Unordered long |
|-------------------------------------------------------|----------------|--------------|------------------|----------------|
| **Existing**                                          | 3390 ms        | 5780 ms      | 3350 ms          | 5420 ms        |
| **Rebase times using first date**                     | 12.7 ms          | 2510 ms      | 12.9 ms          | 1450 ms         |
| **Rebase times using `median_low` date**              | 12.6 ms        | 1270 ms      | 13.1 ms          | 1170 ms         |
| **Iterate using relative timedeltas without sorting** | 14.4 ms        | 15.7 ms      | 15.3 ms            | 1550 ms         |
| **Iterate using relative timedeltas with sorting**    | 15.1 ms        | 16.3 ms      | 16.4 ms          | 15 ms        |

Rebasing the times either using the first date of the array or the median date works well for arrays with narrow time ranges, but does not do as well for arrays that span a long range of times.  Not surprisingly, sorting is required to get the best performance with the iterative relative timedelta approach in the case of an unordered long time range array.  Note even in the sorting case I am careful to restore the original order of the times, so that is factored into the performance as well.

In summary, I think the iterative approach that sorts and then restores the order of the original dates provides consistently good performance, though it is not necessarily the fastest approach for every use case.  If this all makes sense, I can try and sort out the edge cases and put together a cleaned up PR that implements this optimized approach.

cc: @dcherian

(Updated the "Unordered long" case to cover a similar span of time to the "Ordered long" case -- previously it did span as much time).

**Comment by jswhit** [comment]
Wow, those are some pretty impressive speed-ups!  Please do submit a PR.

**Comment by spencerkclark** [comment]
Great!  See #270 for the PR.

**Comment by spencerkclark** [comment]
Closed by #270.
